src/common/lpspi/bus.rs

```diff
@@ -0,0 +1,344 @@
+
+        self.clear_fifos();
+
+        let byteorder = sequence.byteorder;
+
+        let read_part = &mut self.read_part;
+        let write_part = &mut self.write_part;
+
+        let (mut write_dma, mut read_dma) = match &mut self.dma {
+            LpspiDma::Disabled => (None, None),
+            LpspiDma::Partial(dma) => match sequence.recommended_dma_direction() {
+                TransferDirection::Read => (None, Some(dma)),
+                TransferDirection::Write => (Some(dma), None),
+            },
+            LpspiDma::Full(dma1, dma2) => (Some(dma1), Some(dma2)),
+        };
+
+        let read_task = async {
+            if let Some(phase1) = &sequence.phase1 {
+                let actions = phase1.get_read_actions();
+                read_part.perform_read_actions(actions, byteorder).await;
+            }
+            if let Some(phase2) = &sequence.phase2 {
+                if let Some(actions) = phase2.get_read_actions() {
+                    read_part.perform_read_actions(actions, byteorder).await;
+                }
+            }
+        };
+        let write_task = async {
+            let has_phase_1 = sequence.phase1.is_some();
+            let has_phase_2 = sequence.phase2.is_some();
+
+            if let Some(phase1) = &sequence.phase1 {
+                let actions = phase1.get_write_actions();
+                write_part
+                    .perform_write_actions(
+                        actions,
+                        false,
+                        has_phase_2,
+                        byteorder,
+                        write_dma.as_deref_mut(),
+                    )
+                    .await;
+            }
+            if let Some(phase2) = &sequence.phase2 {
+                let actions = phase2.get_write_actions();
+                write_part
+                    .perform_write_actions(
+                        actions,
+                        has_phase_1,
+                        false,
+                        byteorder,
+                        write_dma.as_deref_mut(),
+                    )
+                    .await;
+            }
+        };
+
+        futures::join!(read_task, write_task);
+
+        self.check_errors()
+    }
+
+    /// Perform a sequence of transfer actions while continuously checking for errors.
+    async unsafe fn transfer(&mut self, sequence: ActionSequence<'_>) -> Result<(), LpspiError> {
+        let mut cleanup_on_error = CleanupOnError::new(self);
+        let this = cleanup_on_error.driver();
+
+        let data = this.data;
+
+        let result: Result<(), LpspiError> = futures::select_biased! {
+            res = data.lpspi.watch_for_errors().fuse() => res,
+            res = this.transfer_unchecked(sequence).fuse() => res,
+        };
+
+        cleanup_on_error.finish(result)
+    }
+
+    /// Returns whether or not the busy flag is set.
+    fn busy(&self) -> bool {
+        ral::read_reg!(ral::lpspi, self.lpspi(), SR, MBF == MBF_1)
```

Comment:
Turns out this is buggy on imxrt1062 (or all, maybe)? Writing to the FIFO does **not** set `MBF` immediately; if compiled with optimizations this has a chance to report `not busy` although a transfer is in progress.